File: ios/App/Repository/ProcessingRepo.swift
//
//  This file is part of Blokada.
//
//  This Source Code Form is subject to the terms of the Mozilla Public
//  License, v. 2.0. If a copy of the MPL was not distributed with this
//  file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
//  Copyright © 2021 Blocka AB. All rights reserved.
//
//  @author Karol Gusak
//

import Foundation
import Combine

// This repo is used by all other repos to report their processing state globally.
class ProcessingRepo: Startable {

    var errorsHot: AnyPublisher<ComponentError, Never> {
        self.writeError.compactMap { $0 }.eraseToAnyPublisher()
    }

    var ongoingHot: AnyPublisher<ComponentOngoing, Never> {
        self.writeOngoing.compactMap { $0 }.eraseToAnyPublisher()
    }

    var currentlyOngoingHot: AnyPublisher<[ComponentOngoing], Never> {
        self.ongoingHot.scan([]) { acc, it in
            if it.ongoing && !acc.contains(where: { $0.component == it.component }) {
                return acc + [it]
            } else if !it.ongoing && acc.contains(where: { $0.component == it.component }) {
                return acc.filter { $0.component != it.component }
            } else {
                return acc
            }
        }
        .removeDuplicates()
        .eraseToAnyPublisher()
    }

    fileprivate let writeError = CurrentValueSubject<ComponentError?, Never>(nil)
    fileprivate let writeOngoing = CurrentValueSubject<ComponentOngoing?, Never>(nil)

    // Subscribers with lifetime same as the repository
    private var cancellables = Set<AnyCancellable>()

    func start() {
    }

    func notify(_ component: Any, _ error: Error, major: Bool) {
        writeError.send(ComponentError(
            component: String(describing: component), error: error, major: major
        ))
        notify(component, ongoing: false)
    }

    func notify(_ component: Any, ongoing: Bool) {
        writeOngoing.send(ComponentOngoing(
            component: String(describing: component), ongoing: ongoing
        ))
    }

}

class DebugProcessingRepo: ProcessingRepo {

    private let log = BlockaLogger("Processing")
    private var cancellables = Set<AnyCancellable>()

    override func start() {
        super.start()

        errorsHot.sink(
            onValue: { it in
                if it.major {
                    self.log.e("Major error: \(it.component): \(it.error)")
                } else {
                    self.log.w("Error: \(it.component): \(it.error)")
                }
            }
        )
        .store(in: &cancellables)

        currentlyOngoingHot.sink(
            onValue: { it in
                self.log.v("\(it)")
            }
        )
        .store(in: &cancellables)

    }

    func injectOngoing(_ ongoing: ComponentOngoing) {
        writeOngoing.send(ongoing)
    }

}

File: android5/app/src/main/java/repository/ProcessingRepo.kt
/*
 * This file is part of Blokada.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * Copyright © 2022 Blocka AB. All rights reserved.
 *
 * @author Karol Gusak (karol@blocka.net)
 */

package repository

import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import model.*
import utils.Logger

open class ProcessingRepo {

    private val writeOngoing = MutableSharedFlow<ComponentOngoing?>()
    private val writeError = MutableSharedFlow<ComponentError?>()
    internal val writeConnIssues = MutableSharedFlow<Set<Any>>(replay = 1)

    private val ongoingHot = writeOngoing.filterNotNull().distinctUntilChanged()
    val errorsHot = writeError.filterNotNull().distinctUntilChanged()
    val connIssuesHot = writeConnIssues.map { it.isNotEmpty() }.distinctUntilChanged()

    internal val recentTimeoutsHot = writeError.scan(emptyList<ComponentTimeout>()) { acc, p ->
        if (p == null) {
            acc
        } else if (p.error !is TimeoutException) {
            // Remove old timeouts
            acc.filter { it.timeoutMillis + 30 * 1000 >= System.currentTimeMillis() }
        } else if (acc.any { it.component == p.component }) {
            // Replace timeout of recently timed-out task with a newer timestamp
            acc.filter { it.component != p.component }.plus(
                ComponentTimeout(p.component, System.currentTimeMillis())
            )
        } else {
            // Add this task to timeouts timestamps
            acc + ComponentTimeout(p.component, System.currentTimeMillis())
        }
    }.distinctUntilChanged()

    val currentlyOngoingHot = ongoingHot.scan(emptyList<ComponentOngoing>()) { acc, p ->
        if (p.ongoing && !acc.any { it.component == p.component }) {
            acc + listOf(p)
        } else if (!p.ongoing && acc.any { it.component == p.component }) {
            acc.filter { it.component != p.component }
        } else {
            acc
        }
    }.distinctUntilChanged()

    open fun start() {
        GlobalScope.launch { writeConnIssues.emit(emptySet()) }
        onManyRecentTimeouts_ReportConnIssue()
    }

    suspend fun notify(component: Any, ex: BlokadaException, major: Boolean) {
        writeError.emit(ComponentError("$component", ex, major))
        notify(component, ongoing = false)
    }

    suspend fun notify(component: Any, ongoing: Boolean) {
        writeOngoing.emit(ComponentOngoing("$component", ongoing))
    }

    suspend fun reportConnIssues(component: Any, experiencing: Boolean) {
        val current = writeConnIssues.first()
        if (experiencing) writeConnIssues.emit(current + component)
        else writeConnIssues.emit(current - component)
    }

    private fun onManyRecentTimeouts_ReportConnIssue() {
        GlobalScope.launch {
            recentTimeoutsHot.collect {
                reportConnIssues("timeout", it.size > 3)
            }
        }
    }

}

class DebugProcessingRepo: ProcessingRepo() {

    override fun start() {
        super.start()

        GlobalScope.launch {
            errorsHot.collect {
                if (it.major)
                    Logger.e("Processing", "Major error: ${it.component}: ${it.error}")
                else
                    Logger.w("Processing", "Error: ${it.component}: ${it.error}")
            }
        }

        GlobalScope.launch {
            currentlyOngoingHot.collect {
                Logger.v("Processing", "$it")
            }
        }

        GlobalScope.launch {
            writeConnIssues.distinctUntilChanged().collect {
                Logger.v("Processing", "ConnIssues: $it")
            }
        }

        GlobalScope.launch {
            recentTimeoutsHot.distinctUntilChanged().collect {
                Logger.v("Processing", "Timeout: $it")
            }
        }
    }

}